hls-test-utils/src/Test/Hls.hs

```diff
@@ -134,3 +137,15 @@ runSessionWithServer' plugin conf sconf caps root s = withLock lock $ keepCurren
       (t, _) <- duration $ cancel server
       putStrLn $ "Finishing canceling (took " <> showDuration t <> "s)"
   pure x
+
+-- | Wait for all progress to be done
+-- Needs at least one progress done notification to return
+waitForProgressDone :: Session ()
+waitForProgressDone = loop
+  where
+    loop = do
```

Comment:
`whileM` from extra can be used to do this - you just return True to keep going.

Reply:
The codebase consistenly uses recursive functions for looping this way, so switching to `whileM` would be inconsistent.

Reply:
Fair enough